File: Add Code Here/C++/Dynamic Programming/Matrix_Chain_Multiplication.cpp
#include <bits/stdc++.h>
using namespace std;
 

int MatrixChainOrder(int p[], int i, int j)
{
    if (i == j)
        return 0;
    int k;
    int mini = INT_MAX;
    int count;
    for (k = i; k < j; k++)
    {
        count = MatrixChainOrder(p, i, k)
                + MatrixChainOrder(p, k + 1, j)
                + p[i - 1] * p[k] * p[j];
        mini = min(count, mini);
    }
    // Return minimum count
    return mini;
}
int main()
{
    int arr[] = { 1, 2, 3, 4, 3 };
    int N = sizeof(arr) / sizeof(arr[0]);
    
    cout << "Minimum number of multiplications is "<< MatrixChainOrder(arr, 1, N - 1);
    return 0;
}

File: Add Code Here/PYTHON/Matrix _chain_multiplication.py
# Matrix chain multiplication using recursion

import sys
# Matrix A[i] has dimension p[i-1] x p[i]
# for i = 1..n
def MatrixChainOrder(p, i, j):
	if i == j:
		return 0
	_min = sys.maxsize
	# Place parenthesis at different places
	# between first and last matrix, 
	# recursively calculate count of multiplications 
	# for each parenthesis placement 
	# and return the minimum count
	for k in range(i, j):
		count = (MatrixChainOrder(p, i, k)
				+ MatrixChainOrder(p, k + 1, j)
				+ p[i-1] * p[k] * p[j])
		if count < _min:
			_min = count
	# Return minimum count
	return _min
# Driver code
if __name__ == '__main__':
	arr = [1, 2, 3, 4, 3]
	N = len(arr)
	# Function call
	print("Minimum number of multiplications is ",
	MatrixChainOrder(arr, 1, N-1))
# Code is contributed by Aviral Srivastava
